simple/reprovide.go

```diff
@@ -49,44 +56,60 @@ func NewReprovider(ctx context.Context, reprovideIniterval time.Duration, rsys r
 
 // Close the reprovider
 func (rp *Reprovider) Close() error {
+	rp.cancel()
+	<-rp.doneCh
 	return nil
 }
 
 // Run re-provides keys with 'tick' interval or when triggered
 func (rp *Reprovider) Run() {
-	// dont reprovide immediately.
-	// may have just started the daemon and shutting it down immediately.
-	// probability( up another minute | uptime ) increases with uptime.
-	after := time.After(time.Minute)
-	var done doneFunc
-	for {
-		if rp.tick == 0 {
-			after = make(chan time.Time)
+	defer close(rp.doneCh)
+
+	var initialReprovideCh, reprovideCh <-chan time.Time
+
+	// If reproviding is enabled (non-zero)
+	if rp.tick > 0 {
+		reprovideTicker := time.NewTicker(rp.tick)
+		defer reprovideTicker.Stop()
+		reprovideCh = reprovideTicker.C
+
+		// If the reprovide ticker is larger than a minute (likely),
+		// provide once after we've been up a minute.
+		//
+		// Don't provide _immediately_ as we might be just about to stop.
+		if rp.tick > time.Minute {
+			initialReprovideTimer := time.NewTimer(time.Minute)
+			defer initialReprovideTimer.Stop()
+
+			initialReprovideCh = initialReprovideTimer.C
 		}
+	}
 
+	var done chan<- error
+	for rp.ctx.Err() == nil {
 		select {
```

Comment:
Technically, if someone tries to reprovide before we get here, we could miss the reprovide trigger. I'm fine with that.